packages/toolkit/src/combineSlices.ts

```diff
@@ -0,0 +1,203 @@
+> = Id<
+  // TODO: use PreloadedState generic instead
+  CombinedState<
+    StaticState & WithRequiredProp<Partial<LazyLoadedState>, InjectedKeys>
+  >
+>
+
+// Prevent undeclared keys in reducer maps
+type ValidateReducerMaps<
+  LazyLoadedState extends Record<string, unknown>,
+  Slices extends [
+    LazyLoadedSlice<LazyLoadedState> | LazyLoadedReducerMap<LazyLoadedState>,
+    ...Array<
+      LazyLoadedSlice<LazyLoadedState> | LazyLoadedReducerMap<LazyLoadedState>
+    >
+  ]
+> = Slices &
+  {
+    [Index in keyof Slices]: Slices[Index] extends AnySlice
+      ? {}
+      : {
+          [Name in keyof Slices[Index]]: Name extends keyof LazyLoadedState
+            ? Reducer
+            : never
+        }
+  }
+
+type NewKeys<
+  LazyLoadedState extends Record<string, unknown>,
+  Slices extends [
+    LazyLoadedSlice<LazyLoadedState> | LazyLoadedReducerMap<LazyLoadedState>,
+    ...Array<
+      LazyLoadedSlice<LazyLoadedState> | LazyLoadedReducerMap<LazyLoadedState>
+    >
+  ]
+> = Slices[number] extends infer Slice
+  ? Slice extends AnySlice
+    ? SliceName<Slice>
+    : keyof Slice
+  : never
+
+interface CombinedSliceReducer<
+  StaticState,
+  LazyLoadedState extends Record<string, unknown> = {},
+  InjectedKeys extends keyof LazyLoadedState = never
+> extends Reducer<
+    CombinedSliceState<StaticState, LazyLoadedState, InjectedKeys>,
+    AnyAction
+  > {
+  withLazyLoadedSlices<
+    Lazy extends Record<string, unknown> = {}
+  >(): CombinedSliceReducer<StaticState, LazyLoadedState & Lazy, InjectedKeys>
+
+  injectSlices<
+    Slices extends [
+      LazyLoadedSlice<LazyLoadedState> | LazyLoadedReducerMap<LazyLoadedState>,
+      ...Array<
+        LazyLoadedSlice<LazyLoadedState> | LazyLoadedReducerMap<LazyLoadedState>
+      >
+    ]
+  >(
+    ...slices: ValidateReducerMaps<LazyLoadedState, Slices>
+  ): CombinedSliceReducer<
+    StaticState,
+    LazyLoadedState,
+    InjectedKeys | NewKeys<LazyLoadedState, Slices>
+  >
+
+  // TODO: deal with nested state?
+  selector<
+    Selected,
+    State extends CombinedSliceState<
+      StaticState,
+      LazyLoadedState,
+      InjectedKeys
+    >,
+    Args extends any[]
+  >(
+    selectorFn: (state: State, ...args: Args) => Selected
+  ): (state: WithOptionalProp<State, InjectedKeys>, ...args: Args) => Selected
+}
+
+type StaticState<
+  Slices extends [AnySlice | ReducerMap, ...Array<AnySlice | ReducerMap>]
+> = UnionToIntersection<
+  Slices[number] extends infer Slice
+    ? Slice extends AnySlice
+      ? WithSlice<Slice>
+      : StateFromReducersMapObject<Slice>
+    : never
+>
+
+const isSlice = (maybeSlice: AnySlice | ReducerMap): maybeSlice is AnySlice =>
+  typeof maybeSlice.actions === 'object'
+
+export function combineSlices<
+  Slices extends [AnySlice | ReducerMap, ...Array<AnySlice | ReducerMap>]
+>(...slices: Slices): CombinedSliceReducer<Id<StaticState<Slices>>> {
+  const reducerMap = slices.reduce<Record<string, Reducer>>((map, slice) => {
```

Comment:
Can I pretty please get a `for of` loop here instead of a `.reduce`? 
Generally, I would love it if we could keep `reduce` at a minimum, and `.forEach` out of the project 😅

Reply:
Or, maybe a use of `flatMap` 🤔 

```js
const getReducers = (slices) => slices.flatMap(sliceOrObject => isSlice(slice) ? [slice.name, slice.reducer] : Object.entries(slice)

const combineSlices = (slices) => Object.from(getReducers (slices))

//further down
const injectSlices = (slices) => {
  for (const [name, reducer] of getReducers (slices)) {
    // inlined injectReducer function
  }
} 
```